File: dsp11.c
#include <graph.h>
#include <conio.h>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define IDIM     320
#define FDIM     320

float   fI[IDIM + 1], fF[FDIM + 1];

void main (void)
    {
    /*
     *** PROGRAM NO.11  CHARACTERISTICS OF 2ND-ORDER SYSTEMS ***
     */
    int    iCount;
    float  fR, fTHETA, fT, fW, fGAIN, fMax, fA, fB, fC, fD;

    _clearscreen  (_GCLEARSCREEN);

    printf ("PROGRAM 11: ");
    printf ("Characteristics of 2nd-order systems\n");

    /*
     * fI=IMPULSE RESPONSE, fF=FREQUENCY RESPONSE (MAGNITUDE) **
     */
    printf ("Enter RADIUS, ANGLE (degrees) of z-plane poles :");
    scanf ("%f, %f", &fR, &fTHETA);

    fT = fTHETA * 3.141593 / 180.0;

    /*
     ***** LOAD IMPULSE RESPONSE, STARTING AT LOCATION fI[2] ***
     */
    fI[1] = 0.0;
    fI[2] = 1.0;

    for (iCount = 3; iCount <= IDIM; iCount++)
        fI[iCount] = 2.0 * fR * cos (fT) * fI[iCount-1] -
                     fR * fR * fI[iCount-2];

    /*
     ******* NORMALISE TO PEAK VALUE OF UNITY, THEN PLOT *******
     */
    fMax = 0.0;

    for (iCount = 1; iCount <= IDIM; iCount++)
        if (fabs (fI[iCount]) > fMax)
            fMax = fabs (fI[iCount]);

    for (iCount = 1; iCount <= IDIM; iCount++)
        fI[iCount] /= fMax;

    _setvideomode( _MRES16COLOR );

    for (iCount = 1; iCount <= IDIM; iCount++)
        {
        _moveto (iCount, 70);
        _lineto (iCount, (int)(70 - fI[iCount] * 30.0 + 0.5));
        }

    /*
     *****************  ESTIMATE PEAK GAIN  ********************
     */
    fA = 1.0 - 2.0 * fR * cos (fT) * cos (fT) +
          fR * fR * cos (2.0 * fT);
    fB = 2.0 * fR * cos (fT) * sin (fT) -
          fR * fR * sin (2.0 * fT);
    fGAIN = 1.0 / pow (fA * fA + fB * fB, 0.5);

    /*
     ****** FIND FREQUENCY RESPONSE (NORMALISED), THEN PLOT ****
     */
    for (iCount = 1; iCount <= IDIM; iCount++)
        {
        fW = (float)(iCount - 1) * 3.141593 / (float)IDIM;
        fC = 1.0 - 2.0 * fR * cos (fT) * cos (fW) +
              fR * fR * cos (2.0 * fW);
        fD = 2.0 * fR * cos (fT) * sin (fW) -
              fR * fR * sin (2.0 * fW);
        fF[iCount] = 1.0/(pow(fC * fC + fD * fD, 0.5) * fGAIN);
        }

    for (iCount = 1; iCount <= IDIM; iCount++)
        {
        _moveto (iCount, 170);
        _lineto (iCount, (int)(170 - fF[iCount] * 60.0 + 0.5));
        }

    getch();
    _setvideomode( _DEFAULTMODE );

    /*
     ***********************************************************
     */
    }

File: dsp11.pas
(****  PROGRAM NO.11  CHARACTERISTICS OF 2ND-ORDER SYSTEMS  ****)
program DSP11;
uses graph3;
var I: array[1..320] of real;
var F: array[1..320] of real;
var R,THETA,T,A,B,GAIN,C,D,W,M: real;
var N: integer;
(***  I=IMPULSE RESPONSE, F=FREQUENCY RESPONSE (MAGNITUDE)  ****)
begin
  writeln('ENTER RADIUS, ANGLE (DEGREES) OF Z-PLANE POLES ');
  readln(R,THETA);
  T:=THETA*3.141593/180;
(******  LOAD IMPULSE RESPONSE, STARTING AT LOCATION I(2)  *****)
  I[1]:=0; I[2]:=1;
  for N:=3 TO 320 do
    begin
      I[N]:=2*R*cos(T)*I[N-1]-R*R*I[N-2];
    end;
(*******  NORMALISE TO PEAK VALUE OF UNITY, THEN  PLOT  ********)
  M:=0;
  for N:=1 to 320 do
    begin
      if abs(I[N])>M then M:=abs(I[N]);
    end;
  for N:=1 to 320 do
    begin
      I[N]:=I[N]/M;
    end;
  graphmode;
  for N:=1 to 320 do
    begin
      draw(N,70,N,round(70-I[N]*30),1);
    end;
(*********************  ESTIMATE PEAK GAIN  ********************)
  A:=1-2*R*cos(T)*cos(T)+R*R*cos(2*T);
  B:=2*R*cos(T)*sin(T)-R*R*sin(2*T);
  GAIN:=1/sqrt(A*A+B*B);
(******  FIND FREQUENCY RESPONSE (NORMALISED), THEN PLOT  ******)
  for N:=1 to 320 do
    begin
      W:=(N-1)*3.141593/320;
      C:=1-2*R*cos(T)*cos(W)+R*R*cos(2*W);
      D:=2*R*cos(T)*sin(W)-R*R*sin(2*W);
      F[N]:=1/(sqrt(C*C+D*D)*GAIN);
    end;
  for N:=1 to 320 do
    begin
      draw(N,170,N,round(170-F[N]*60),1);
    end;
end.
(***************************************************************)